tools/codesigningtool/codesigningtool.py

```diff
@@ -120,16 +122,122 @@ def _certificate_fingerprint(identity):
   fingerprint = fingerprint.replace(":", "")
   return fingerprint
 
+def _certificate_common_name(cert):
+  _, subject, _ = execute.execute_and_filter_output([
+    "openssl",
+    "x509",
+    "-noout",
+    "-inform",
+    "DER",
+    "-subject"
+  ], inputstr=cert, raise_on_failure=True)
+  subject = subject.strip().split('/')
+  cert_cn = [f for f in subject if "CN=" in f]
+
+  if len(cert_cn) == 0:
+    return None
+
+  cert_cn = cert_cn[0]
+  cert_cn = cert_cn.replace("CN=", "")
+
+  return cert_cn
 
 def _get_identities_from_provisioning_profile(mpf):
   """Iterates through all the identities in a provisioning profile, lazily."""
   for identity in mpf["DeveloperCertificates"]:
-    if not isinstance(identity, bytes):
+    cert = _certificate_data(identity)
+    yield _certificate_fingerprint(cert)
+
+def _certificate_data(cert):
+  if not isinstance(cert, bytes):
       # Old versions of plistlib return the deprecated plistlib.Data type
       # instead of bytes.
-      identity = identity.data
-    yield _certificate_fingerprint(identity)
+      cert = cert.data
+
+  return cert
+
+def _get_smartcard_tokens(xml):
+  """Get available tokens from the output of 'system_profiler SPSmartCardsDataType -xml'"""
+  tokens = [x for x in xml if x.get("_name", None) == "AVAIL_SMARTCARDS_TOKEN"]
 
+  if len(tokens) == 0:
+    return []
+
+  tokens = tokens[0].get("_items", None)
+  tokens = [x.get("_name", None) for x in tokens]
+
+  return tokens
+
+def _get_smartcard_keychain(xml):
+  """Get keychain items from the output of 'system_profiler SPSmartCardsDataType -xml'"""
+  keychain = [x for x in xml if x.get("_name", None) == "AVAIL_SMARTCARDS_KEYCHAIN"]
+
+  if len(keychain) == 0:
+    return []
+
+  keychain = keychain[0].get("_items", None)
+  return keychain
+
+def _find_smartcard_identities(identity=None):
+  """Finds smartcard identitites on the current system."""
+  ids = []
+  _, xml, _ = execute.execute_and_filter_output([
+      "/usr/sbin/system_profiler",
+      "SPSmartCardsDataType",
+      "-xml"
+  ], raise_on_failure=True)
```

Comment:
What could cause a failure here? I'm mainly trying to think of cases where this new functionality could cause failures in builds that didn't have them before.

Reply:
I don't think I've ever actually seen `system_profiler` fail unless the datatype happened to be of some type that doesn't yet exist. I'd gone and checked that `SPSmartCardsDataType` is present at least as far back as macOS 10.14.

Reply:
This is just me copy pasting from the other similar invocations 🤦 I don't think we should expect any failure here. So we don't need `raise_on_failure=True`

Reply:
I'm not saying it shouldn't raise (it probably should), I was just wondering in which cases it could fail.

Reply:
That makes sense! Was about to say that it's probably good to keep it even though we haven't found any failures from this command